core/shared/src/test/scala/fs2/concurrent/ChannelSuite.scala

```diff
@@ -251,4 +251,79 @@ class ChannelSuite extends Fs2Suite {
 
     TestControl.executeEmbed(test)
   }
+
+  def checkIfCloseWithElementClosing(channel: IO[Channel[IO, Int]]) =
+    channel.flatMap { ch =>
+      ch.closeWithElement(0) *> ch.isClosed.assert
+    }
+
+  test("closeWithElement closes right after sending the last element in bounded(1) case") {
+    val channel = Channel.bounded[IO, Int](1)
+
+    checkIfCloseWithElementClosing(channel)
+  }
+
+  test("closeWithElement closes right after sending the last element in bounded(2) case") {
+    val channel = Channel.bounded[IO, Int](2)
+
+    checkIfCloseWithElementClosing(channel)
+  }
+
+  test("closeWithElement closes right after sending the last element in unbounded case") {
+    val channel = Channel.unbounded[IO, Int]
+
+    checkIfCloseWithElementClosing(channel)
+  }
+
+  test("closeWithElement closes right after sending the last element in synchronous case") {
+    val channel = Channel.synchronous[IO, Int]
+
+    checkIfCloseWithElementClosing(channel)
+  }
+
+  def racingSendOperations(channel: IO[Channel[IO, Int]]) = {
+    val expectedFirstCase = ((Right(()), Right(())), List(0, 1))
+    val expectedSecondCase = ((Right(()), Left(Channel.Closed)), List(1))
+    val expectedThirdCase = ((Left(Channel.Closed), Right(())), List(1))
+
+    channel
+      .flatMap { ch =>
+        ch.send(0).both(ch.closeWithElement(1)).parProduct(ch.stream.compile.toList)
+      }
+      .map {
+        case obtained @ ((Right(()), Right(())), List(0, 1)) =>
+          assertEquals(obtained, expectedFirstCase)
+        case obtained @ ((Right(()), Left(Channel.Closed)), List(1)) =>
+          assertEquals(obtained, expectedSecondCase)
```

Comment:
Hmm, I don't think this case should ever happen? Is it happening?

Reply:
It is happening once or twice during 1000 runs, but it is pretty rare.

Reply:
I don't see how that can happen if everything is working correctly. If the `closeWithElement` returns that the `Channel` is already closed, then who is closing it? Seems like a bug 😕

Reply:
To be honest, I left this match case there. When we iterated over multiple possible solutions. So, probably this will not happen. I will try it without this match case soon.

Reply:
Yep, I tried with 100000 runs, and that match case never occurred. So, it can be safely removed. 👍